### PR TITLE
Feature/devise

### PR DESCRIPTION
**Pull request by kamran-khalid-shah**
First PR

### CODE REVIEW

File: app/controllers/registrations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
+class RegistrationsController < Devise::RegistrationsController
+  def new
+    @user = User.new
+    @user.build_company
+  end
+
+  def create
+    if User.create_user_and_company(sign_up_params)
```
Comment (bk-az) [inline app/controllers/registrations_controller.rb:8]: can we use `User.create(sign_up_params)`?

File: app/controllers/registrations_controller.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+      flash[:notice] = 'Account created succesfully!'
```
Comment (bk-az) [inline app/controllers/registrations_controller.rb:9]: set owner id of company

File: app/controllers/home_controller.rb [hunk 3]
```diff
@@ -0,0 +1,4 @@
+class HomeController < ApplicationController
+  def index
```
Comment (abdullah-siddiqui-git) [inline app/controllers/home_controller.rb:2]: respond_to do |format|
  format.html
end

Always prefer to add format block

File: app/controllers/registrations_controller.rb (hunk 2 above)
Comment (abdullah-siddiqui-git) [inline app/controllers/registrations_controller.rb]: Use ```I18n``` syntax

File: app/views/devise/mailer/email_changed.html.erb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<p>Hello <%= @email %>!</p>
+
+<% if @resource.try(:unconfirmed_email?) %>
+  <p>We're contacting you to notify you that your email is being changed to <%= @resource.unconfirmed_email %>.</p>
```
Comment (abdullah-siddiqui-git) [inline app/views/devise/mailer/email_changed.html.erb]: use I18n

File: Gemfile [hunk 5]
```diff
@@ -17,20 +17,17 @@ gem 'webpacker', '~> 4.0'
...
 # Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder
 gem 'jbuilder', '~> 2.7'
-# Use Redis adapter to run Action Cable in production
-# gem 'redis', '~> 4.0'
-# Use Active Model has_secure_password
-# gem 'bcrypt', '~> 3.1.7'
 
-# Use Active Storage variant
-# gem 'image_processing', '~> 1.2'
+gem 'devise'
```
Comment (bk-az) [inline Gemfile]: gem 'devise', 'version'

File: app/controllers/registrations_controller.rb [hunk 6]
```diff
@@ -0,0 +1,37 @@
+class RegistrationsController < Devise::RegistrationsController
+  def new
```
Comment (bk-az) [inline app/controllers/registrations_controller.rb:2]: respond_to only html for now

File: app/controllers/registrations_controller.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    end
+  end
+
+  private
+
+  def sign_up_params
+    params.require(:user).permit(:first_name, :last_name, :email, :password, :password_confirmation, company_attributes: [:name, :subdomain])
+  end
+
+  def account_update_params
```
Comment (bk-az) [inline app/controllers/registrations_controller.rb]: remove

File: app/models/user.rb [hunk 8]
```diff
@@ -2,10 +2,15 @@
...
 # Users model
 class User < ApplicationRecord
-  belongs_to :company
-  belongs_to :department
-  belongs_to :team
-  has_many :users_team, dependent: :destroy
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
```
Comment (bk-az) [inline app/models/user.rb:8]: confirmable

File: db/migrate/20200907114306_create_companies.rb [hunk 9]
```diff
@@ -4,10 +4,10 @@
 class CreateCompanies < ActiveRecord::Migration[6.0]
   def change
     create_table :companies do |t|
-      t.string :status, null: false
-      t.string :name, null: false
+      t.string :status
+      t.string :name, null: false, uniqueness: true
```
Comment (bk-az) [inline db/migrate/20200907114306_create_companies.rb]: remove uniqueness: true

File: db/migrate/20200910132642_add_devise_to_users.rb [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+      ## Trackable
+      # t.integer  :sign_in_count, default: 0, null: false
+      # t.datetime :current_sign_in_at
+      # t.datetime :last_sign_in_at
+      # t.string   :current_sign_in_ip
+      # t.string   :last_sign_in_ip
+
+      ## Confirmable
+      # t.string   :confirmation_token
```
Comment (bk-az) [inline db/migrate/20200910132642_add_devise_to_users.rb]: uncomment this

File: db/migrate/20200910132642_add_devise_to_users.rb [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+      # t.datetime :locked_at
+
+
+      # Uncomment below if timestamps were not included in your original model.
+      # t.timestamps null: false
+    end
+
+    add_index :users, :email,                unique: true
+    add_index :users, :reset_password_token, unique: true
+    # add_index :users, :confirmation_token,   unique: true
```
Comment (bk-az) [inline db/migrate/20200910132642_add_devise_to_users.rb]: uncomment

File: app/controllers/registrations_controller.rb [hunk 12]
```diff
@@ -0,0 +1,39 @@
+class RegistrationsController < Devise::RegistrationsController
+  def new
+    respond_to do |format|
```
Comment (bk-az) [inline app/controllers/registrations_controller.rb:5]: respond_to block should be at the end of action

**Pull request merged**